zilliqa/src/node.rs

```diff
@@ -263,26 +263,66 @@ impl Node {
         self.consensus.head_block().header.number
     }
 
-    pub fn resolve_block_number(&self, number: BlockNumberOrTag) -> u64 {
-        // TODO(#863): This doesn't respect two-chain finalization.
-        match number {
-            BlockNumberOrTag::Latest => self.get_chain_tip(),
-            BlockNumberOrTag::Finalized => self.get_chain_tip().saturating_sub(2),
-            BlockNumberOrTag::Safe => self.get_chain_tip().saturating_sub(2),
+    pub fn resolve_block_number(&self, block_number: BlockNumberOrTag) -> u64 {
+        match block_number {
+            BlockNumberOrTag::Number(n) => n,
             BlockNumberOrTag::Earliest => 0,
+            BlockNumberOrTag::Latest => self.get_chain_tip(),
             BlockNumberOrTag::Pending => self.get_chain_tip(),
-            BlockNumberOrTag::Number(n) => n,
+            BlockNumberOrTag::Finalized => {
+                let Ok(Some(view)) = self.db.get_latest_finalized_view() else {
+                    return 0u64;
+                };
+                let Ok(Some(block)) = self.db.get_block_by_view(&view) else {
+                    return 0u64;
+                };
+                block.number()
+            }
+            // Safe block tag in our consensus refers to the block that the node's highQC points to
+            // (high_qc means it's the latest = high, and it's a QC where 2/3 validators voted for it).
+            BlockNumberOrTag::Safe => {
+                let block_hash = self.consensus.high_qc.block_hash;
+
+                let Ok(Some(safe_block)) = self.consensus.get_block(&block_hash) else {
+                    return 0u64;
+                };
+                safe_block.number()
+            }
         }
     }
 
     pub fn get_block(&self, block_id: impl Into<BlockId>) -> Result<Option<Block>> {
         match block_id.into() {
             BlockId::Hash(RpcBlockHash {
                 block_hash,
-                require_canonical: _,
+                require_canonical,
             }) => {
-                // TODO(#863): require_canonical
-                self.consensus.get_block(&block_hash.into())
+                // See https://eips.ethereum.org/EIPS/eip-1898
+                let Ok(Some(block)) = self.consensus.get_block(&block_hash.into()) else {
+                    const RESOURCE_NOT_FOUND: i32 = -32001;
```

Comment:
I don't think these eth-API level concerns should be happening in `node.rs`. Why not return `None` in this case and map to the error code in the eth API impls?